lightning-invoice/src/payment.rs

```diff
@@ -562,22 +630,26 @@ where
 				} else if *rejected_by_dest {
 					log_trace!(self.logger, "Payment {} rejected by destination; not retrying", log_bytes!(payment_hash.0));
 					self.payer.abandon_payment(payment_id.unwrap());
+					self.remove_path_inflight_htlcs(path);
 				} else if retry.is_none() {
 					log_trace!(self.logger, "Payment {} missing retry params; not retrying", log_bytes!(payment_hash.0));
 					self.payer.abandon_payment(payment_id.unwrap());
+					self.remove_path_inflight_htlcs(path);
 				} else if self.retry_payment(payment_id.unwrap(), *payment_hash, retry.as_ref().unwrap()).is_ok() {
 					// We retried at least somewhat, don't provide the PaymentPathFailed event to the user.
 					return;
 				} else {
 					self.payer.abandon_payment(payment_id.unwrap());
+					self.remove_path_inflight_htlcs(path);
 				}
 			},
 			Event::PaymentFailed { payment_hash, .. } => {
 				self.remove_cached_payment(&payment_hash);
 			},
 			Event::PaymentPathSuccessful { path, .. } => {
-				let path = path.iter().collect::<Vec<_>>();
-				self.scorer.lock().payment_path_successful(&path);
+				let path_with_refs = path.iter().collect::<Vec<_>>();
+				self.scorer.lock().payment_path_successful(&path_with_refs);
+				self.remove_path_inflight_htlcs(path);
```

Comment:
got it. if we track amount and path info in `payment_cache,` is there a way to key the path information with `PaymentHash` uniquely? or can we guarantee that it will not try the same path again?

I'm thinking specifically of what happens when we re-attempt to send an HTLC through a path that was attempted before and still outstanding. 

I think we decided to track inflight htlcs in a separate map so that they are easier to be looked up by scid across all payments, rather than local to them, which is similar to what we do [here](https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/routing/router.rs#L1563-L1566) in the router.

Reply:
> > I think we decided to track inflight htlcs in a separate map so that they are easier to be looked up by scid across all payments, rather than local to them, which is similar to what we do [here](https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/routing/router.rs#L1563-L1566) in the router.
> 
> Yea, both approaches are valid, I just worry about cases like the one you describe where we could accidentally have too many paths being tracked in the map and we end up forever thinking there's paths pending. I don't have specific examples, however, so I could just be wrong.

In worse case, I think we'd subtract too much and so would need a `saturating_sub`. i.e., The map always starts with zero, but there can be in-flight HTLCs from prior to a restart. And any HTLCs sent after a restart would have been added already.

I think the larger point, though, is when calling `channel_penalty_msat` we'd prefer to do one map lookup. Otherwise, we'd have to scan all the paths stored in every entry of `payment_cache` in order to sum up the in-flight HTLCs (i.e., a channel can have an in-flight HTLC for more than one payment, IIUC).

Reply:
Ah, that makes sense. So basically, store paths in `payment_cache`. Then when needing to call `find_route`, produce the map from `payment_cache`, lock the scorer, wrap it and the map in an internally defined scorer, and pass that to `find_route`.

Reply:
thanks for chiming in @jkczyz @TheBlueMatt -- this does make more sense now.

just to make sure I understand correctly: it sounds like we are okay trading off the efficiency of a single map lookup so we do not have to keep the data within `inflight_htlcs_map` in sync with `payment_cache`? When routing, we'll just simply compute the same contract (`HashMap<(u64, bool), u64)>`) from `payment_cache` to `find_route`?

Reply:
Yeah, sorta. We still get the benefit of a single map look-up in `find_route` each time it calls `channel_penalty_msat`. It's just that we compute that map (i.e., `inflight_htlcs_map` once prior to each call to `find_route`. The map itself will actually be pretty small. (i.e., it will contain less channels than channels considered by `find_route`).

Will look something like this:

```rust
struct AccountForInFlightHtlcs<'a, S: Score> {
    scorer: &'a S,
    inflight_htlcs: HashMap<(u64, bool), u64)>,
}

impl<'_, S: Score> Score for AccountForInFlightHtlcs<S, '_> {
    // ...
}

let inflight_htlcs = ...;
let locked_scorer = self.scorer.lock();
let scorer = AccountForInFlightHtlcs { scorer: &*locked_scorer, inflight_htlcs };
let route = self.router.find_route(... , &scorer);
```